### ISSUE TITLE
Primitive literals in computed property names

### ISSUE THREAD

**Issue by weswigham** [Microsoft/TypeScript#4653]
Looking at #4648 brought this up -
```ts
declare module JSX {
  interface IntrinsicElements {
    ["my-custom-element"]: MyCustomElementClass
  }
}
```

We should allow string literals in computed property names in interfaces. Right now we only allow well-know symbols, but allowing _all_ primitive literals (of valid indexer types, I suppose) should be supported.

This is a spec change - section 2.2.3 would need to be updated to reflect this, but, IMO, is fairly important and worth the change. We can't strongly type, say, `tslint.json` result structures right now, as we lack this.

**Comment by danquirk** [comment]
See https://github.com/Microsoft/TypeScript/issues/1082 for some previous discussion on computed property names in interfaces.

**Comment by vladima** [comment]
will this work in your case?
```ts
declare module JSX {
    interface IntrinsicElements {
        "my-element": __React.HTMLAttributes
    }
}

<my-element autoPlay/>
```

**Comment by weswigham** [comment]
Huh. Why does our syntax for defining non-identifier string object properties in interfaces not overlap with the computed property syntax?

That's weird and unexpected.

**Comment by vladima** [comment]
I guess that it reflects syntax for defining object literals. Also string\numeric literals as names in properties were allowed in TypeScript long before ES6 support was added

**Comment by weswigham** [comment]
Weird - VSCode's syntax highlighter doesn't even think it's valid:
![image](https://cloud.githubusercontent.com/assets/2932786/9695610/61ab922c-5318-11e5-8d90-ce5e3ce24921.png)

Like, I feel like the square brackets (computed property) should certainly be correct syntax in TS, even if the old interface syntax is still valid - just because it works for symbols, so it feels like it aught to work for other primitives.

**Comment by weswigham** [comment]
I also don't think many people would argue with me that 
![image](https://cloud.githubusercontent.com/assets/2932786/9695764/2d63dce8-531a-11e5-81af-dcfe70366ee0.png)
is better written as
![image](https://cloud.githubusercontent.com/assets/2932786/9695771/3fb4838e-531a-11e5-8f64-76752b3a8a9f.png)

The top is feels almost nonsensical (I did not even realize that was valid TS until I tried it), the bottom is ES6 classes (and yet invalid in TS).

**Comment by danquirk** [comment]
Both of those forms are valid in classes (where someone would actually migrate JS to TS). Would be reasonable to support both in interfaces for consistency sake.

**Comment by jbondc** [comment]
Also was confused by this recently. Consider #2225
![image](https://cloud.githubusercontent.com/assets/199519/10074458/12666f76-629e-11e5-8b34-b2baeb46f781.png)

Don't really know if string literals should be in computed property names but you should be able to see those strings somewhere.

**Comment by DanielRosenwasser** [comment]
@jbondc I'm not really sure how #2225 ties in, but #606 is relevant.

**Comment by weswigham** [comment]
@DanielRosenwasser in the chrome devtools, one can write
```js
obj.t/**/
```
get completions at the marker, and recieve the list of properties
```
thing
thing-explainer
$thing
other-thing
```
select `thing-explainer`, and have it completed as
```js
obj["thing-explainer"]/**/
```

I think this is the behavior we want to be able to replicate.

**Comment by vladima** [comment]
should be addressed in #5535
